### PR TITLE
Implement W-buffer and fix depth-mapping

### PR DESCRIPTION
**Pull request by JayFoxRox**
This PR [implements W-Buffers](https://www.mvps.org/directx/articles/using_w-buffers.htm) and fixes hardware depth-mapping (scale + offset).
The formulas for the depth-mapping were guessed after testing on hardware.

The results are still about 3 bits inprecise in the lower bits due to rounding and precision issues probably. This is still somewhat better than the current `master` solution which produces complete garbage for some mappings.

This PR implements this in the SW-Rasterizer and the HW-Renderer.

---

This PR was [tested using a HW-Test](https://github.com/JayFoxRox/3ds-tests/tree/master/depth-write) ([Planar results where top.zw = bottom.zw](https://docs.google.com/spreadsheets/d/1E8xvJIbj_MVygWtr4ELPuzOJPqEv4yGeHQuKG6kFmTc/edit#gid=0) & [Perspective results](https://docs.google.com/spreadsheets/d/13_m98KE29fwTgUowAM43-eh9nLw7l2TEXwtYVhclEpA/edit#gid=0)).
There are no known regressions while this might fix some games.

The interpolations weren't tested much yet (not at all for depth-scale and offset) but it's assumed this will work. More tests could be done, feel free to add them to the sheets I linked.

(I'd like to thank everyone who tested including but not limited to @MerryMage @JohnGodGames and @LokiX)

---

~~This fixes the depth formulas but not depth clearing / buffer access. There will be another PR soon (almost complete) to fix the other bugs caused by #1462 .~~ #1624

### CODE REVIEW

File: src/video_core/rasterizer.cpp [hunk 1]
```diff
@@ -858,10 +858,31 @@ static void ProcessTriangleInternal(const Shader::OutputVertex& v0,
...
+            float depth_offset = float24::FromRaw(regs.viewport_depth_near_plane).ToFloat32();
+            float depth = interpolated_z_over_w * depth_scale + depth_offset;
+
+            // Potentially switch to W-Buffer
+            if (regs.depthmap_enable == Pica::Regs::DepthBuffering::WBuffering) {
+                float interpolated_w = interpolated_w_inverse.ToFloat32() * wsum;
+
+                // W-Buffer (z * scale + w * offset = (z / w * scale + offset) * w)
+                depth *= interpolated_w;
+
```
Comment (lioncash) [inline src/video_core/rasterizer.cpp]: Unnecessary newline. Other than that, no real complaints from me.

**Comment by MerryMage** [comment]
> The results are still about 3 bits inprecise in the lower bits due to rounding and precision issues probably. 

There should probably be a comment about this in the code so that future-people know about this (perhaps someone would want/need to have increased accuracy in the future).

File: src/video_core/rasterizer.cpp [hunk 2]
```diff
@@ -858,10 +858,31 @@ static void ProcessTriangleInternal(const Shader::OutputVertex& v0,
...
+            float depth_scale = float24::FromRaw(regs.viewport_depth_range).ToFloat32();
+            float depth_offset = float24::FromRaw(regs.viewport_depth_near_plane).ToFloat32();
+            float depth = interpolated_z_over_w * depth_scale + depth_offset;
+
+            // Potentially switch to W-Buffer
+            if (regs.depthmap_enable == Pica::Regs::DepthBuffering::WBuffering) {
+                float interpolated_w = interpolated_w_inverse.ToFloat32() * wsum;
+
+                // W-Buffer (z * scale + w * offset = (z / w * scale + offset) * w)
+                depth *= interpolated_w;
```
Comment (JayFoxRox) [inline src/video_core/rasterizer.cpp]: I believe W (upper case) is 1/w but I'm not sure.
I'll remove the interpolated_w variable for clarity and merge it with this line because I'm not sure about the terminology and the extra variable is not necessary. The current name is clearly a misnomer.
(I don't want another interpolated_w_inverse2 variable either)

File: src/video_core/rasterizer.cpp [hunk 3]
```diff
@@ -990,6 +1011,7 @@ static void ProcessTriangleInternal(const Shader::OutputVertex& v0,
                         return 255 - combiner_output.a();
 
                     case Regs::BlendFactor::DestAlpha:
+                    case Regs::BlendFactor::DestColor:
```
Comment (Subv) [inline src/video_core/rasterizer.cpp]: Was this change intended?

**Comment by wwylele** [comment]
It seems there is memory leak. When I test this PR, Citra continues eating up memory to several GBs when running. I don't know if this is a bug in PR, or a hidden bug in master.

**Comment by KillzXGaming** [comment]
Must be PR. Just tried both latest, and this merged and this uses 4-5gb compared to 300mb in master.

File: src/video_core/rasterizer.cpp (hunk 3 above)
Comment (JayFoxRox) [inline src/video_core/rasterizer.cpp]: Nope, I'll remove it. Should probably be patched elsewhere

**Comment by JayFoxRox** [comment]
I've addressed the comments.

I used the style for the inprecise message roughly from elsewhere in the same file (without `// TODO`) - there are far more precision issues I've seen by now so I don't want to stress it enough yet anyway.

I also removed the newline despite feeling the old style looked more relaxed.

I also tried reproducing the memory leak and I've tested this branch using "gcc version 5.3.0 (GCC)" and "clang version 3.7.1 (tags/RELEASE_371/final)", 4.4.5-1-ARCH Linux and Mesa 11.1.2 on Ivybridge HD4000.
I tested a handful of games and never got to leak more than ~10kB (all from fontconfig = not my code) and memory usage was well below 400MB even with debug builds enabled.

What's your configuration (System, compiler, game, ...) for the leak situation and are you testing this on a clean master?

**Comment by KillzXGaming** [comment]
Ok. Got the latest master, then merged it, this is only PR I merged.
Specs:
GTX 970
8GB RAM
I5 4670 (non K)
Full Info on specs
http://pastebin.com/e0EVuVUV

**Comment by JayFoxRox** [comment]
I need to know which game etc. I don't really care about the hardware.
I'm starting to think this is related to failed shader compilation which creates a giant log in memory.
Does anyone have a log or is able to apitrace it to see what goes wrong (probably on nvidia drivers)?

**Comment by KillzXGaming** [comment]
Edit: Looking at Leo626 software render fixes it for me aswell. Also fixes glitches from animal crossing.

**Comment by wwylele** [comment]
I test this PR on Windows (MSVC). All of my bootable game (totally 9 games) experience such memory consuming on every frame since opening. But the consuming speed are not the same. The worst one among them is Kirby Triple Deluxe. On current master all of them are fine. 
I don't know how to use apitrace...

**Comment by wwylele** [comment]
I apitrace the first ten seconds running Kirby Triple Deluxe. [Here](https://www.dropbox.com/s/fsnvdgaj9kaj9mm/citra-qt_1621.trace?dl=0) is the trace file.

File: src/video_core/renderer_opengl/gl_rasterizer.h [hunk 4]
```diff
@@ -141,6 +143,8 @@ struct PicaShaderConfig {
         return std::memcmp(this, &o, sizeof(PicaShaderConfig)) == 0;
     };
 
+    Pica::Regs::DepthBuffering depthmap_enable = Pica::Regs::DepthBuffering::WBuffering;
+
```
Comment (wwylele) [inline src/video_core/renderer_opengl/gl_rasterizer.h]: Found a weird stuff here. I tried adding `u32 padding;` here, and the memory leak mentioned above was surprisingly fixed! 
I did this after noticing there follows a 64-bit hash function for `struct PicaShaderConfig` and thinking something about memory alignment.

Comment (wwylele) [inline src/video_core/renderer_opengl/gl_rasterizer.h]: On a second thought, the 'memory leak' is caused by the buggy implement of operator== and hash for `struct PicaShaderConfig`, because there can be padding in the struct with undetermined value.

Comment (lioncash) [inline src/video_core/renderer_opengl/gl_rasterizer.h]: Just putting this here for notification purposes:

As said on IRC in the developer channel, I'll take care of this.

**Comment by JayFoxRox** [comment]
So, memory hog aside (by adding padding temporarily for testing this PR in case you run into problems), please test if this PR works properly.
In case you find remaining depth-differences in SW/HW or regressions please get on IRC and `/query` me from \#citra-dev instead of filling up this thread (there might be a lot of false positives).
Thanks

**Comment by JohnGodgames** [comment]
Recent master vs w-buffer

**Brave Tank Hero**
brave citra vs w-buffer

**Crash Time 3D**
crash time 3d citra vs w-buffer

**Dr. Mario Miracle Cure**
dr mario citra vs w-buffer

**Face Racers - Photo Finish**
face racers citra vs w-buffer

and propably more...

**Comment by amirlou** [comment]
hello Jay this fixed Rune Factory 4 but after i merged this with Hardware Renderer Texture Forwarding and wanted to build it in appveyo i got this error its work fine with current master
 C:\projects\citra\src\video_core\renderer_opengl\gl_rasterizer.cpp(263): error C2039: 'shader_dirty': is not a member of 'OpenGLState::<unnamed-type-draw>' [C:\projects\citra\build\src\video_core\video_core.vcxproj]
fixed

**Comment by amirlou** [comment]
Rune Factory 4 
fixed

File: src/video_core/renderer_opengl/gl_rasterizer.h (hunk 4 above)
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_rasterizer.h]: Give this [branch](https://github.com/lioncash/citra/tree/pica-config) a try. I'll likely resolve the hashing to be member-wise and not just glob the whole struct itself.

**Comment by bunnei** [comment]
Code LGTM, just a few nits on comments/codestyle.

**Comment by amirlou** [comment]
hello Jay when i merged this and Hardware Renderer Texture Forwarding i got this error in AppVeyor

(ClCompile target) -> 
C:\projects\citra\src\video_core\renderer_opengl\gl_rasterizer.cpp(265): error C2039: 'shader_dirty': is not a member of 'OpenGLState::<unnamed-type-draw>' [C:\projects\citra\build\src\video_core\video_core.vcxproj]

**Comment by JayFoxRox** [comment]
(rebased onto master to keep local and remote in sync - we still need a solution for the comparion bug so I'll re-test rebased version if there is a solution [I'm not satisfied by lioncash's current proposal])

File: src/video_core/renderer_opengl/gl_rasterizer.cpp [hunk 5]
```diff
@@ -865,9 +870,9 @@ void RasterizerOpenGL::SyncCullMode() {
 
 void RasterizerOpenGL::SyncDepthModifiers() {
     float depth_scale = -Pica::float24::FromRaw(Pica::g_state.regs.viewport_depth_range).ToFloat32();
```
Comment (JayFoxRox) [inline src/video_core/renderer_opengl/gl_rasterizer.cpp]: I'll change this so it doesn't have a negative depth_scale here. It's pretty annoying for fog and creates a difference between HW and SW logic which might confuse some people

File: src/video_core/renderer_opengl/gl_shader_gen.cpp [hunk 6]
```diff
@@ -566,7 +567,15 @@ vec4 secondary_fragment_color = vec4(0.0);
     }
 
     out += "color = last_tex_env_out;\n";
-    out += "gl_FragDepth = gl_FragCoord.z + depth_offset;\n}";
+
+    out += "float z_over_w = gl_FragCoord.z * 2.0 - 1.0;\n";
```
Comment (JayFoxRox) [inline src/video_core/renderer_opengl/gl_shader_gen.cpp]: This will be inverted soon too

**Comment by bunnei** [comment]
@JayFoxRox can you please post an update here on the status of this change?

**Comment by JayFoxRox** [comment]
Waiting for #1704 , I have a rebased version with comments addressed locally and will push it once #1704 is merged.

**Comment by JayFoxRox** [comment]
@wwylele: does it work for you without leak now?

I rebased this onto the new master (post #1704) and also fixed the things I mentioned in the diff comments above.

I played OoT for 10 minutes and it only used about 5MB more after that (which is the shader cache *slowly* filling). Also I tested some depth-write tests and they still work fine :)
- So if this doesn't leak/trash the memory anymore this should be ready to be merged.

**Comment by wwylele** [comment]
@JayFoxRox Yeah now it works fine without leak!

**Comment by Subv** [comment]
LGTM
